Description: Fix the compilation errors with Java 8
Author: Emmanuel Bourg <ebourg@apache.org>
Forwarded: no
--- a/src/jalview/ws/rest/InputType.java
+++ b/src/jalview/ws/rest/InputType.java
@@ -42,7 +42,6 @@
 import org.apache.http.entity.mime.content.ContentBody;
 import org.apache.http.entity.mime.content.StringBody;
 
-import sun.io.CharacterEncoding;
 import sun.misc.CharacterEncoder;
 
 /***
@@ -263,4 +262,4 @@
     return new Option(name,descr, req, defType==null ? "" : defType.toString(), curType==null ? "" : curType.toString(),molType.toStringValues(),
             null);
   }
-}
\ No newline at end of file
+}
