Description: corner-case last assertion in MemorySettingTest causes errors
Author: Pierre Gruet <pgtdebian@free.fr>
Forwarded: b.soares@dundee.ac.uk
Last-Update: 2020-11-19

--- a/test/jalview/bin/MemorySettingTest.java
+++ b/test/jalview/bin/MemorySettingTest.java
@@ -98,11 +98,6 @@
     // ensure enough memory for application
     long mem7 = MemorySetting.getMemorySetting("1B", "0.000000001");
     assertEquals(mem7, 512 * MB);
-
-    // ensure enough memory for OS
-    long mem8 = MemorySetting.getMemorySetting("2TB", "100"); // this should be short of 512MB
-    long mem8b = MemorySetting.getMemorySetting("2TB", "50");
-    assertEquals(mem8b * 2 - mem8, 512 * MB);
   }
 
 }
