<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (Version 2.7)
 * Copyright (C) 2011 J Procter, AM Waterhouse, G Barton, M Clamp, S Searle
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * 
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
-->
<head>
<title>Release History</title>
</head>
<body>
<p><strong>Release History</strong></p>
<table border="1">
	<tr>
		<td width="60" nowrap>
		<div align="center"><em><strong>Release</strong></em></div>
		</td>
		<td>
		<div align="center"><em><strong>New Features</strong></em></div>
		</td>
		<td>
		<div align="center"><em><strong>Issues Resolved</strong></em></div>
		</td>
	</tr>
		<tr>
			<td>
				<div align="center">
					<strong><a name="Jalview2.7">2.7</a>
					</strong><br> <em>27/09/2011</em>
				</div></td>
			<td><em>Application</em>
				<ul>
					<li>Jalview Desktop News Reader</li>
					<li>Tweaked default layout of web services menu</li>
					<li>View/alignment association menu to enable user to easily
						specify which alignment a multi-structure view takes its
						colours/correspondences from</li>
					<li>Allow properties file location to be specified as URL </li>
						<li>Extend jalview project to preserve associations between
							many alignment views and a single Jmol display</li>
						<li>Store annotation row height in jalview project file</li>
						<li>Annotation row column label formatting attributes stored
							in project file</li>
						<li>Annotation row order for auto-calculated annotation rows
							preserved in jalview project file</li>
						<li>Visual progress indication when Jalview state is saved
							using Desktop window menu</li>
						<li>Visual indication that command line arguments are still
							being processed</li>
						<li>Groovy script execution from URL</li>
						<li>Colour by annotation default min and max colours in
							preferences</li>
						<li>Automatically associate PDB files dragged onto an
							alignment with sequences that have high similarity and matching
							IDs</li>
						<li>Update JGoogleAnalytics to latest release (0.3)</li>
						<li>&#39;view structures&#39; option to open many structures
							in same window</li>
						<li>Sort associated views menu option for tree panel</li>
						<li>Group all JABA and non-JABA services for a particular
							analysis function in its own submenu</li>
				</ul> <em>Applet</em>
			<ul>
					<li>Userdefined and autogenerated annotation rows for groups</li>
					<li>Adjustment of alignment annotation pane height</li>
					<li>Annotation scrollbar for annotation panel</li>
					<li>Drag to reorder annotation rows in annotation panel</li>
					<li>&#39;automaticScrolling&#39; parameter</li>
					<li>Allow sequences with partial ID string matches to be
						annotated from GFF/jalview features files</li>
					<li>Sequence logo annotation row in applet</li>
					<li>Absolute paths relative to host server in applet
						parameters are treated as such</li>
					<li>New in the JalviewLite javascript API:
						<ul>
							<li>JalviewLite.js javascript library</li>
							<li>Javascript callbacks for
								<ul>
									<li>Applet initialisation</li>
									<li>Sequence/alignment mouse-overs and selections</li>
								</ul>
							</li>
							<li>scrollTo row and column alignment scrolling functions</li>
							<li>Select sequence/alignment regions from javascript</li>
							<li>javascript structure viewer harness to pass messages
								between Jmol and Jalview when running as distinct applets</li>
							<li>sortBy method</li>
							<li>Set of applet and application examples shipped with
								documentation</li>
							<li>New example to demonstrate jalviewLite and Jmol
								javascript message exchange</li>
						</ul>
				</ul> <em>General</em>
				<ul>
					<li>Enable Jmol displays to be associated with multiple
						multiple alignments</li>
					<li>Option to automatically sort alignment with new tree</li>
					<li>User configurable link to enable redirects to a
						www.jalview.org mirror</li>
					<li>Jmol colours option for Jmol displays</li>
					<li>Configurable newline string when writing alignment and
						other flat files</li>
					<li>Allow alignment annotation description lines to contain
						html tags</li>
				</ul> <em>Documentation and Development</em>
				<ul>
					<li>Add groovy test harness for bulk load testing to examples
					</li>
					<li>Groovy script to load and align a set of sequences using a
						web service before displaying the result in the Jalview desktop</li>
					<li>Restructured javascript and applet api documentation</li>
					<li>Ant target to publish example html files with applet archive</li>
					<li>Netbeans project for building jalview from source</li>
					<li>ant task to create online javadoc for jalview source</li>
				</ul></td>
			<td><em>Application</em>
				<ul>
					<li>User defined colourscheme throws exception when current
						built in colourscheme is saved as new scheme</li>
					<li>AlignFrame-&gt;Save in application pops up save dialog for
						valid filename/format</li>
					<li>Cannot view associated structure for Uniprot sequence</li>
					<li>PDB file association breaks for Uniprot sequence P37173</li>
					<li>Associate PDB from file dialog does not tell you which
						sequence is to be associated with the file</li>
					<li>Find All raises null pointer exception when query only
						matches sequence IDs</li>
					<li>Pre 2.6 Jalview project cannot be loaded into v2.6</li>
					<li>Jalview project with Jmol views created with Jalview 2.4
						cannot be loaded</li>
					<li>Filetype associations not installed for webstart launch</li>
					<li>Two or more chains in a single PDB file associated with
						sequences in different alignments do not get coloured by their
						associated sequence</li>
					<li>Visibility status of autocalculated annotation row not
						preserved when project is loaded</li>
					<li>Annotation row height and visibility attributes not stored
						in jalview project</li>
					<li>Tree bootstraps are not preserved when saved as a jalview
						project</li>
					<li>Envision2 workflow tooltips are corrupted</li>
					<li>Enabling show group conservation also enables colour by
						conservation</li>
					<li>Duplicate group associated conservation or consensus
						created on new view</li>
					<li>Annotation scrollbar not displayed after &#39;show all
						hidden annotation rows&#39; option selected</li>
					<li>Alignment quality not updated after alignment annotation
						row is hidden then shown</li>
					<li>Preserve colouring of structures coloured by sequences in pre
						jalview 2.7 projects</li>
					<li>Web service job parameter dialog is not laid out properly
					</li>
					<li>Web services menu not refreshed after &#39;reset
						services&#39; button is pressed in preferences</li>
					<li>Annotation off by one in jalview v2_3 example project</li>
					<li>Structures imported from file and saved in project get
						name like jalview_pdb1234.txt when reloaded</li>
					<li>Jalview does not always retrieve progress of a JABAWS job
						execution in full once it is complete</li>
				</ul> <em>Applet</em>
				<ul>
					<li>Alignment height set incorrectly when lots of annotation
						rows are displayed</li>
					<li>Relative URLs in feature HTML text not resolved to
						codebase</li>
					<li>View follows highlighting does not work for positions in
						sequences</li>
					<li>&lt;= shown as = in tooltip</li>
					<li>Export features raises exception when no features exist</li>
					<li>Separator string used for serialising lists of IDs for
						javascript api is modified when separator string provided as
						parameter</li>
					<li>Null pointer exception when selecting tree leaves for
						alignment with no existing selection</li>
					<li>Relative URLs for datasources assumed to be relative to
						applet&#39;s codebase</li>
					<li>Status bar not updated after finished searching and search
						wraps around to first result</li>
					<li>StructureSelectionManager instance shared between several
						jalview applets causes race conditions and memory leaks</li>
					<li>Hover tooltip and mouseover of position on structure not
						sent from Jmol in applet</li>
					<li>Certain sequences of javascript method calls to applet API
						fatally hang browser</li>
				</ul> <em>General</em>
				<ul>
					<li>View follows structure mouseover scrolls beyond position
						with wrapped view and hidden regions</li>
					<li>Find sequence position moves to wrong residue with/without
						hidden columns</li>
					<li>Sequence length given in alignment properties window is
						off by 1</li>
					<li>InvalidNumberFormat exceptions thrown when trying to
						import PDB like structure files</li>
					<li>Positional search results are only highlighted between
						user-supplied sequence start/end bounds</li>
					<li>End attribute of sequence is not validated</li>
					<li>Find dialog only finds first sequence containing a given
						sequence position</li>
					<li>Sequence numbering not preserved in MSF alignment output</li>
					<li>Jalview PDB file reader does not extract sequence from
						nucleotide chains correctly</li>
					<li>Structure colours not updated when tree partition changed
						in alignment</li>
					<li>Sequence associated secondary structure not correctly
						parsed in interleaved stockholm</li>
					<li>Colour by annotation dialog does not restore current state
					</li>
					<li>Hiding (nearly) all sequences doesn&#39;t work properly</li>
					<li>Sequences containing lowercase letters are not properly
						associated with their pdb files</li>
				</ul> <em>Documentation and Development</em>
				<ul>
					<li>schemas/JalviewWsParamSet.xsd corrupted by ApplyCopyright
						tool</li>
				</ul></td>
		</tr>
		<tr>
			<td>
				<div align="center">
					<strong><a name="Jalview2.6.1">2.6.1</a>
					</strong><br> <em>15/11/2010</em>
				</div></td>
			<td><em>Application</em>
				<ul>
					<li>New warning dialog when the Jalview Desktop cannot contact
						web services</li>
					<li>JABA service parameters for a preset are shown in service
						job window</li>
					<li>JABA Service menu entries reworded</li>
				</ul></td>
			<td>
				<ul>
					<li>Modeller PIR IO broken - cannot correctly import a pir
						file emitted by jalview</li>
					<li>Existing feature settings transferred to new alignment
						view created from cut'n'paste</li>
					<li>Improved test for mixed amino/nucleotide chains when
						parsing PDB files</li>
					<li>Consensus and conservation annotation rows occasionally
						become blank for all new windows</li>
					<li>Exception raised when right clicking above sequences in
						wrapped view mode</li>
				</ul> <em>Application</em>
				<ul>
					<li>multiple multiply aligned structure views cause cpu usage
						to hit 100% and computer to hang</li>
					<li>Web Service parameter layout breaks for long user
						parameter names</li>
					<li>Jaba service discovery hangs desktop if Jaba server is
						down</li>
				</ul></td>
		</tr>
		<tr>
		<td>
		<div align="center"><strong><a name="Jalview2.6">2.6</a></strong><br>
		<em>26/9/2010</em></div>
		</td>
		<td><em>Application</em>
		<ul>
			<li>Support for <strong>Ja</strong>va <strong>b</strong>ioinformatics
			<strong>a</strong>nalysis <strong>w</strong>eb <strong>s</strong>ervices (JABAWS)</li>
			<li>Web Services preference tab</li>
			<li>Analysis parameters dialog box and user defined preferences</li>
			<li>Improved speed and layout of Envision2 service menu</li>
			<li>Superpose structures using associated sequence alignment</li>
			<li>Export coordinates and projection as CSV from PCA viewer</li>
		</ul>
		<em>Applet</em>
		<ul>
			<li>enable javascript: execution by the applet via the link out
			mechanism</li>
		</ul>
		<em>Other</em>
		<ul>
			<li>Updated the Jmol Jalview interface to work with Jmol series
			12</li>
			<li>The Jalview Desktop and JalviewLite applet now require Java
			1.5</li>
			<li>Allow Jalview feature colour specification for GFF sequence
			annotation files</li>
			<li>New 'colour by label' keword in jalview feature file type
			colour specification</li>
			<li>New Jalview Desktop Groovy API method that allows a script
			to check if it being run in an interactive session or in a batch
			operation from the jalview command line</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>clustalx colourscheme colours Ds preferentially when both
			D+E are present in over 50% of the column</li>
		</ul>

		<em>Application</em>
		<ul>
			<li>typo in AlignmentFrame-&gt;View-&gt;Hide-&gt;all but
			selected Regions menu item</li>
			<li>sequence fetcher replaces ',' for ';' when the ',' is part
			of a valid accession ID</li>
			<li>fatal OOM if object retrieved by sequence fetcher runs out
			of memory</li>
			<li>unhandled Out of Memory Error when viewing pca analysis
			results</li>
			<li>InstallAnywhere builds fail to launch on OS X java 10.5
			update 4 (due to apple Java 1.6 update)</li>
			<li>Installanywhere Jalview silently fails to launch</li>
		</ul>
		<em>Applet</em>
		<ul>
			<li>Jalview.getFeatureGroups() raises an
			ArrayIndexOutOfBoundsException if no feature groups are defined.</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong><a name="Jalview2.5.1">2.5.1</a></strong><br>
		<em>14/6/2010</em></div>
		</td>
		<td></td>
		<td>
		<ul>
			<li>Alignment prettyprinter doesn't cope with long sequence IDs
			</li>
			<li>clustalx colourscheme colours Ds preferentially when both
			D+E are present in over 50% of the column</li>
			<li>nucleic acid structures retrieved from PDB do not import
			correctly</li>
			<li>More columns get selected than were clicked on when a number
			of columns are hidden</li>
			<li>annotation label popup menu not providing correct
			add/hide/show options when rows are hidden or none are present</li>
			<li>Stockholm format shown in list of readable formats, and
			parser copes better with alignments from RFAM.</li>
			<li>CSV output of consensus only includes the percentage of all
			symbols if sequence logo display is enabled</li>

		</ul>
		<em>Applet</em>
		<ul>
			<li>annotation panel disappears when annotation is
			hidden/removed</li>
		</ul>
		<em>Application</em>
		<ul>
			<li>Alignment view not redrawn properly when new alignment
			opened where annotation panel is visible but no annotations are
			present on alignment</li>
			<li>pasted region containing hidden columns is incorrectly
			displayed in new alignment window</li>
			<li>Jalview slow to complete operations when stdout is flooded
			(fix is to close the Jalview console)</li>
			<li>typo in AlignmentFrame-&gt;View-&gt;Hide-&gt;all but
			selected Rregions menu item.</li>
			<li>inconsistent group submenu and Format submenu entry 'Un' or
			'Non'conserved</li>
			<li>Sequence feature settings are being shared by multiple
			distinct alignments</li>
			<li>group annotation not recreated when tree partition is
			changed</li>
			<li>double click on group annotation to select sequences does
			not propagate to associated trees</li>
			<li>Mac OSX specific issues:
			<ul>
				<li>exception raised when mouse clicked on desktop window
				background</li>
				<li>Desktop menu placed on menu bar and application name set
				correctly</li>
				<li>sequence feature settings not wide enough for the save
				feature colourscheme button</li>
			</ul>
			</li>
		</ul>
		</td>
	</tr>
	<tr>

		<td>
		<div align="center"><strong><a name="Jalview2.5">2.5</a></strong><br>
		<em>30/4/2010</em></div>
		</td>
		<td><em>New Capabilities</em>
		<ul>
			<li>URL links generated from description line for
			regular-expression based URL links (applet and application)
			<li>Non-positional feature URL links are shown in link menu</li>
			<li>Linked viewing of nucleic acid sequences and structures</li>
			<li>Automatic Scrolling option in View menu to display the
			currently highlighted region of an alignment.</li>
			<li>Order an alignment by sequence length, or using the average
			score or total feature count for each sequence.</li>
			<li>Shading features by score or associated description</li>
			<li>Subdivide alignment and groups based on identity of selected
			subsequence (Make Groups from Selection).</li>
			<li>New hide/show options including Shift+Control+H to hide
			everything but the currently selected region.</li>
			<!-- introduced but not yet documented <li>Experimental blast report parser</li> -->
		</ul>
		<em>Application</em>
		<ul>
			<li>Fetch DB References capabilities and UI expanded to support
			retrieval from DAS sequence sources</li>
			<li>Local DAS Sequence sources can be added via the command line
			or via the Add local source dialog box.</li>
			<li>DAS Dbref and DbxRef feature types are parsed as database
			references and protein_name is parsed as description line (BioSapiens
			terms).</li>
			<li>Enable or disable non-positional feature and database
			references in sequence ID tooltip from View menu in application.</li>
			<!--			<li>New hidden columns and rows and representatives capabilities
			in annotations file (in progress - not yet fully implemented)</li> -->
			<li>Group-associated consensus, sequence logos and conservation
			plots</li>
			<li>Symbol distributions for each column can be exported and
			visualized as sequence logos</li>
			<li>Optionally scale multi-character column labels to fit within
			each column of annotation row<!-- todo for applet --></li>
			<li>Optional automatic sort of associated alignment view when a
			new tree is opened.</li>
			<li>Jalview Java Console</li>
			<li>Better placement of desktop window when moving between
			different screens.</li>
			<li>New preference items for sequence ID tooltip and consensus
			annotation</li>
			<li>Client to submit sequences and IDs to <a
				href="webServices/index.html#envision2">Envision2</a> Workflows</li>
			<li><em>Vamsas Capabilities</em>
			<ul>
				<li>Improved VAMSAS synchronization (jalview archive used to
				preserve views, structures, and tree display settings)</li>
				<li>Import of vamsas documents from disk or URL via command
				line</li>
				<li>Sharing of selected regions between views and with other
				VAMSAS applications (Experimental feature!)</li>
				<li>Updated API to VAMSAS version 0.2</li>
			</ul>
			</li>
		</ul>
		<em>Applet</em>
		<ul>
			<li>Middle button resizes annotation row height</li>
			<li>New Parameters
			<ul>
				<li>sortByTree (true/false) - automatically sort the associated
				alignment view by the tree when a new tree is opened.</li>
				<li>showTreeBootstraps (true/false) - show or hide branch
				bootstraps (default is to show them if available)</li>
				<li>showTreeDistances (true/false) - show or hide branch
				lengths (default is to show them if available)</li>
				<li>showUnlinkedTreeNodes (true/false) - indicate if
				unassociated nodes should be highlighted in the tree view</li>
				<li>heightScale and widthScale (1.0 or more) - increase the
				height or width of a cell in the alignment grid relative to the
				current font size.</li>
			</ul>
			</li>
			<li>Non-positional features displayed in sequence ID tooltip</li>
		</ul>
		<em>Other</em>
		<ul>
			<li>Features format: graduated colour definitions and
			specification of feature scores</li>
			<li>Alignment Annotations format: new keywords for group
			associated annotation (GROUP_REF) and annotation row display
			properties (ROW_PROPERTIES)</li>
			<li>XML formats extended to support graduated feature
			colourschemes, group associated annotation, and profile visualization
			settings.</li>
		</td>
		<td>
		<ul>
			<li>Source field in GFF files parsed as feature source rather
			than description</li>
			<li>Non-positional features are now included in sequence feature
			and gff files (controlled via non-positional feature visibility in
			tooltip).</li>
			<li>URL links generated for all feature links (bugfix)</li>
			<li>Added URL embedding instructions to features file
			documentation.</li>
			<li>Codons containing ambiguous nucleotides translated as 'X' in
			peptide product</li>
			<li>Match case switch in find dialog box works for both sequence
			ID and sequence string and query strings do not have to be in upper
			case to match case-insensitively.</li>
			<li>AMSA files only contain first column of multi-character
			column annotation labels</li>
			<li>Jalview Annotation File generation/parsing consistent with
			documentation (e.g. Stockholm annotation can be exported and
			re-imported)</li>
			<li>PDB files without embedded PDB IDs given a friendly name</li>
			<li>Find incrementally searches ID string matches as well as
			subsequence matches, and correctly reports total number of both.</li>
			<li>Application:
			<ul>
				<li>Better handling of exceptions during sequence retrieval</li>
				<li>Dasobert generated non-positional feature URL link text
				excludes the start_end suffix</li>
				<li>DAS feature and source retrieval buttons disabled when
				fetch or registry operations in progress.</li>
				<li>PDB files retrieved from URLs are cached properly</li>
				<li>Sequence description lines properly shared via VAMSAS</li>
				<li>Sequence fetcher fetches multiple records for all data
				sources</li>
				<li>Ensured that command line das feature retrieval completes
				before alignment figures are generated.</li>
				<li>Reduced time taken when opening file browser for first
				time.</li>
				<li>isAligned check prior to calculating tree, PCA or
				submitting an MSA to JNet now excludes hidden sequences.</li>
				<li>User defined group colours properly recovered from Jalview
				projects.</li>
			</ul>
			</li>
		</ul>
		</td>

	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.4.0.b2</strong><br>
		28/10/2009</div>
		</td>
		<td>
		<ul>
			<li>Experimental support for google analytics usage tracking.</li>
			<li>Jalview privacy settings (user preferences and docs).</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Race condition in applet preventing startup in jre1.6.0u12+.</li>
			<li>Exception when feature created from selection beyond length
			of sequence.</li>
			<li>Allow synthetic PDB files to be imported gracefully</li>
			<li>Sequence associated annotation rows associate with all
			sequences with a given id</li>
			<li>Find function matches case-insensitively for sequence ID
			string searches</li>
			<li>Non-standard characters do not cause pairwise alignment to
			fail with exception</li>
		</ul>
		<em>Application Issues</em>
		<ul>
			<li>Sequences are now validated against EMBL database</li>
			<li>Sequence fetcher fetches multiple records for all data
			sources</li>
		</ul>
		<em>InstallAnywhere Issues</em>
		<ul>
			<li>Dock icon works for Mac OS X java (Mac 1.6 update issue with
			installAnywhere mechanism)</li>
			<li>Command line launching of JARs from InstallAnywhere version
			(java class versioning error fixed)</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>

		<div align="center"><strong>2.4</strong><br>
		27/8/2008</div>
		</td>
		<td><em>User Interface</em>
		<ul>
			<li>Linked highlighting of codon and amino acid from translation
			and protein products</li>
			<li>Linked highlighting of structure associated with residue
			mapping to codon position</li>
			<li>Sequence Fetcher provides example accession numbers and
			'clear' button</li>
			<li>MemoryMonitor added as an option under Desktop's Tools menu</li>
			<li>Extract score function to parse whitespace separated numeric
			data in description line</li>
			<li>Column labels in alignment annotation can be centred.</li>
			<li>Tooltip for sequence associated annotation give name of
			sequence</li>
		</ul>
		<em>Web Services and URL fetching</em>
		<ul>
			<li>JPred3 web service</li>
			<li>Prototype sequence search client (no public services
			available yet)</li>
			<li>Fetch either seed alignment or full alignment from PFAM</li>
			<li>URL Links created for matching database cross references as
			well as sequence ID</li>
			<li>URL Links can be created using regular-expressions</li>
		</ul>
		<em>Sequence Database Connectivity</em>
		<ul>
			<li>Retrieval of cross-referenced sequences from other databases
			</li>
			<li>Generalised database reference retrieval and validation to
			all fetchable databases</li>
			<li>Fetch sequences from DAS sources supporting the sequence
			command</li>
		</ul>
		<em>Import and Export</em>
		<li>export annotation rows as CSV for spreadsheet import</li>
		<li>Jalview projects record alignment dataset associations, EMBL
		products, and cDNA sequence mappings</li>
		<li>Sequence Group colour can be specified in Annotation File</li>
		<li>Ad-hoc colouring of group in Annotation File using RGB
		triplet as name of colourscheme</li>
		</ul>
		<em>VAMSAS Client capabilities (Experimental)</em>
		<ul>
			<li>treenode binding for VAMSAS tree exchange</li>
			<li>local editing and update of sequences in VAMSAS alignments
			(experimental)</li>
			<li>Create new or select existing session to join</li>
			<li>load and save of vamsas documents</li>
		</ul>
		<em>Application command line</em>
		<ul>
			<li>-tree parameter to open trees (introduced for passing from
			applet)</li>
			<li>-fetchfrom command line argument to specify nicknames of DAS
			servers to query for alignment features</li>
			<li>-dasserver command line argument to add new servers that are
			also automatically queried for features</li>
			<li>-groovy command line argument executes a given groovy script
			after all input data has been loaded and parsed</li>
		</ul>
		<em>Applet-Application data exchange</em>
		<ul>
			<li>Trees passed as applet parameters can be passed to
			application (when using &quot;View in full application&quot;)</li>
		</ul>
		<em>Applet Parameters</em>
		<ul>
			<li>feature group display control parameter</li>
			<li>debug parameter</li>
			<li>showbutton parameter</li>
		</ul>
		<em>Applet API methods</em>
		<ul>
			<li>newView public method</li>
			<li>Window (current view) specific get/set public methods</li>
			<li>Feature display control methods</li>
			<li>get list of currently selected sequences</li>
		</ul>
		<em>New Jalview distribution features</em>
		<ul>
			<li>InstallAnywhere Installer upgraded to IA 2008 VP1</li>
			<li>RELEASE file gives build properties for the latest Jalview
			release.</li>
			<li>Java 1.1 Applet build made easier and donotobfuscate
			property controls execution of obfuscator</li>
			<li>Build target for generating source distribution</li>
			<li>Debug flag for javacc</li>
			<li>.jalview_properties file is documented (slightly) in
			jalview.bin.Cache</li>
			<li>Continuous Build Integration for stable and development
			version of Application, Applet and source distribution</li>
		</ul>

		</td>
		<td>
		<ul>
			<li>selected region output includes visible annotations (for
			certain formats)</li>
			<li>edit label/displaychar contains existing label/char for
			editing</li>
			<li>update PDBEntries when DBRefEntries change (vamsas)</li>
			<li>shorter peptide product names from EMBL records</li>
			<li>Newick string generator makes compact representations</li>
			<li>bootstrap values parsed correctly for tree files with
			comments</li>
			<li>pathological filechooser bug avoided by not allowing
			filenames containing a ':'</li>
			<li>Fixed exception when parsing GFF files containing global
			sequence features</li>
			<li>Alignment datasets are finalized only when number of
			references from alignment sequences goes to zero</li>
			<li>Close of tree branch colour box without colour selection
			causes cascading exceptions</li>
			<li>occasional negative imgwidth exceptions</li>
			<li>better reporting of non-fatal warnings to user when file
			parsing fails.</li>
			<li>Save works when Jalview project is default format</li>
			<li>Save as dialog opened if current alignment format is not a
			valid output format</li>
			<li>Uniprot canonical names introduced for both das and vamsas</li>
			<li>Histidine should be midblue (not pink!) in Zappo</li>
			<li>error messages passed up and output when data read fails</li>
			<li>edit undo recovers previous dataset sequence when sequence
			is edited</li>
			<li>allow PDB files without pdb ID HEADER lines (like those
			generated by MODELLER) to be read in properly</li>
			<li>allow reading of JPred concise files as a normal filetype</li>
			<li>Stockholm annotation parsing and alignment properties import
			fixed for PFAM records</li>
			<li>Structure view windows have correct name in Desktop window
			list</li>
			<li>annotation consisting of sequence associated scores can be
			read and written correctly to annotation file</li>
			<li>Aligned cDNA translation to aligned peptide works correctly</li>
			<li>Fixed display of hidden sequence markers and non-italic font
			for representatives in Applet</li>
			<li>Applet Menus are always embedded in applet window on Macs.</li>
			<li>Newly shown features appear at top of stack (in Applet)</li>
			<li>Annotations added via parameter not drawn properly due to
			null pointer exceptions</li>
			<li>Secondary structure lines are drawn starting from first
			column of alignment</li>
			<li>Uniprot XML import updated for new schema release in July
			2008</li>
			<li>Sequence feature to sequence ID match for Features file is
			case-insensitive</li>
			<li>Sequence features read from Features file appended to all
			sequences with matching IDs</li>
			<li>PDB structure coloured correctly for associated views
			containing a sub-sequence</li>
			<li>PDB files can be retrieved by applet from Jar files</li>
			<li>feature and annotation file applet parameters referring to
			different directories are retrieved correctly</li>
			<!--<li>DAS Histogram Features display ( (Prlic et al) </li>-->
			<li>Fixed application hang whilst waiting for splash-screen
			version check to complete</li>
			<li>Applet properly URLencodes input parameter values when
			passing them to the launchApp service</li>
			<li>display name and local features preserved in results
			retrieved from web service</li>
			<li>Visual delay indication for sequence retrieval and sequence
			fetcher initialisation</li>
			<li>updated Application to use DAS 1.53e version of dasobert DAS
			client</li>
			<li>Re-instated Full AMSA support and .amsa file association</li>
			<li>Fixed parsing of JNet Concise annotation <em>sans</em>
			sequences</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.3</strong><br>
		9/5/07</div>
		</td>
		<td>
		<ul>
			<li>Jmol 11.0.2 integration</li>
			<li>PDB views stored in Jalview XML files</li>
			<li>Slide sequences</li>
			<li>Edit sequence in place</li>
			<li>EMBL CDS features</li>
			<li>DAS Feature mapping</li>
			<li>Feature ordering</li>
			<li>Alignment Properties</li>
			<li>Annotation Scores</li>
			<li>Sort by scores</li>
			<li>Feature/annotation editing in applet</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Headless state operation in 2.2.1</li>
			<li>Incorrect and unstable DNA pairwise alignment</li>
			<li>Cut and paste of sequences with annotation</li>
			<li>Feature group display state in XML</li>
			<li>Feature ordering in XML</li>
			<li>blc file iteration selection using filename # suffix</li>
			<li>Stockholm alignment properties</li>
			<li>Stockhom alignment secondary structure annotation</li>
			<li>2.2.1 applet had no feature transparency</li>
			<li>Number pad keys can be used in cursor mode</li>
			<li>Structure Viewer mirror image resolved</li>
		</ul>
		</td>

	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.2.1</strong><br>
		12/2/07</div>
		</td>
		<td>
		<ul>
			<li>Non standard characters can be read and displayed
			<li>Annotations/Features can be imported/exported to the applet
			via textbox
			<li>Applet allows editing of sequence/annotation/group name
			&amp; description
			<li>Preference setting to display sequence name in italics
			<li>Annotation file format extended to allow Sequence_groups to
			be defined
			<li>Default opening of alignment overview panel can be specified
			in preferences
			<li>PDB residue numbering annotation added to associated
			sequences
		</ul>
		</td>
		<td>
		<ul>
			<li>Applet crash under certain Linux OS with Java 1.6 installed
			<li>Annotation file export / import bugs fixed
			<li>PNG / EPS image output bugs fixed
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.2</strong><br>
		27/11/06</div>
		</td>
		<td>
		<ul>
			<li>Multiple views on alignment
			<li>Sequence feature editing
			<li>&quot;Reload&quot; alignment
			<li>&quot;Save&quot; to current filename
			<li>Background dependent text colour
			<li>Right align sequence ids
			<li>User-defined lower case residue colours
			<li>Format Menu
			<li>Select Menu
			<li>Menu item accelerator keys
			<li>Control-V pastes to current alignment
			<li>Cancel button for DAS Feature Fetching
			<li>PCA and PDB Viewers zoom via mouse roller
			<li>User-defined sub-tree colours and sub-tree selection
			<li>'New Window' button on the 'Output to Text box'
		</ul>
		</td>
		<td>
		<ul>
			<li>New memory efficient Undo/Redo System
			<li>Optimised symbol lookups and conservation/consensus
			calculations
			<li>Region Conservation/Consensus recalculated after edits
			<li>Fixed Remove Empty Columns Bug (empty columns at end of
			alignment)
			<li>Slowed DAS Feature Fetching for increased robustness.
			<li>Made angle brackets in ASCII feature descriptions display
			correctly
			<li>Re-instated Zoom function for PCA
			<li>Sequence descriptions conserved in web service analysis
			results
			<li>Uniprot ID discoverer uses any word separated by &#8739;
			<li>WsDbFetch query/result association resolved
			<li>Tree leaf to sequence mapping improved
			<li>Smooth fonts switch moved to FontChooser dialog box.
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.1.1</strong><br>
		12/9/06</div>
		</td>
		<td>
		<ul>
			<li>Copy consensus sequence to clipboard</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Image output - rightmost residues are rendered if sequence
			id panel has been resized</li>
			<li>Image output - all offscreen group boundaries are rendered</li>
			<li>Annotation files with sequence references - all elements in
			file are relative to sequence position</li>
			<li>Mac Applet users can use Alt key for group editing</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.1</strong><br>
		22/8/06</div>
		</td>
		<td>
		<ul>
			<li>MAFFT Multiple Alignment in default Web Service list</li>
			<li>DAS Feature fetching</li>
			<li>Hide sequences and columns</li>
			<li>Export Annotations and Features</li>
			<li>GFF file reading / writing</li>
			<li>Associate structures with sequences from local PDB files</li>
			<li>Add sequences to exisiting alignment</li>
			<li>Recently opened files / URL lists</li>
			<li>Applet can launch the full application</li>
			<li>Applet has transparency for features (Java 1.2 required)</li>
			<li>Applet has user defined colours parameter</li>
			<li>Applet can load sequences from parameter &quot;sequence<em>x</em>&quot;</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Redundancy Panel reinstalled in the Applet</li>
			<li>Monospaced font - EPS / rescaling bug fixed</li>
			<li>Annotation files with sequence references bug fixed</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.08.1</strong><br>
		2/5/06</div>
		</td>
		<td>
		<ul>
			<li>Change case of selected region from Popup menu</li>
			<li>Choose to match case when searching</li>
			<li>Middle mouse button and mouse movement can compress / expand
			the visible width and height of the alignment</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Annotation Panel displays complete JNet results</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.08b</strong><br>
		18/4/06</div>
		</td>
		<td>&nbsp;</td>
		<td>
		<ul>
			<li>Java 1.5 bug - InternalMessageDialog fix for threads</li>
			<li>Righthand label on wrapped alignments shows correct value</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.08</strong><br>
		10/4/06</div>
		</td>
		<td>
		<ul>
			<li>Editing can be locked to the selection area</li>
			<li>Keyboard editing</li>
			<li>Create sequence features from searches</li>
			<li>Precalculated annotations can be loaded onto alignments</li>
			<li>Features file allows grouping of features</li>
			<li>Annotation Colouring scheme added</li>
			<li>Smooth fonts off by default - Faster rendering</li>
			<li>Choose to toggle Autocalculate Consensus On/Off</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Drag &amp; Drop fixed on Linux</li>
			<li>Jalview Archive file faster to load/save, sequence
			descriptions saved.</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.07</strong><br>
		12/12/05</div>
		</td>
		<td>
		<ul>
			<li>PDB Structure Viewer enhanced</li>
			<li>Sequence Feature retrieval and display enhanced</li>
			<li>Choose to output sequence start-end after sequence name for
			file output</li>
			<li>Sequence Fetcher WSDBFetch@EBI</li>
			<li>Applet can read feature files, PDB files and can be used for
			HTML form input</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>HTML output writes groups and features</li>
			<li>Group editing is Control and mouse click</li>
			<li>File IO bugs</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.06</strong><br>
		28/9/05</div>
		</td>
		<td>
		<ul>
			<li>View annotations in wrapped mode</li>
			<li>More options for PCA viewer</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>GUI bugs resolved</li>
			<li>Runs with -nodisplay from command line</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td height="63">
		<div align="center"><strong>2.05b</strong><br>
		15/9/05</div>
		</td>
		<td>
		<ul>
			<li>Choose EPS export as lineart or text</li>
			<li>Jar files are executable</li>
			<li>Can read in Uracil - maps to unknown residue</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Known OutOfMemory errors give warning message</li>
			<li>Overview window calculated more efficiently</li>
			<li>Several GUI bugs resolved</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.05</strong><br>
		30/8/05</div>
		</td>
		<td>
		<ul>
			<li>Edit and annotate in &quot;Wrapped&quot; view</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Several GUI bugs resolved</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.04</strong><br>
		24/8/05</div>
		</td>
		<td>
		<ul>
			<li>Hold down mouse wheel &amp; scroll to change font size</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Improved JPred client reliability</li>
			<li>Improved loading of Jalview files</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.03</strong><br>
		18/8/05</div>
		</td>
		<td>
		<ul>
			<li>Set Proxy server name and port in preferences</li>
			<li>Multiple URL links from sequence ids</li>
			<li>User Defined Colours can have a scheme name and added to
			Colour Menu</li>
			<li>Choose to ignore gaps in consensus calculation</li>
			<li>Unix users can set default web browser</li>
			<li>Runs without GUI for batch processing</li>
			<li>Dynamically generated Web Service Menus</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>InstallAnywhere download for Sparc Solaris</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.02</strong><br>
		18/7/05</div>
		</td>
		<td>&nbsp;</td>
		<td>
		<ul>
			<li>Copy &amp; Paste order of sequences maintains alignment
			order.</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.01</strong><br>
		12/7/05</div>
		</td>
		<td>
		<ul>
			<li>Use delete key for deleting selection.</li>
			<li>Use Mouse wheel to scroll sequences.</li>
			<li>Help file updated to describe how to add alignment
			annotations.</li>
			<li>Version and build date written to build properties file.</li>
			<li>InstallAnywhere installation will check for updates at
			launch of Jalview.</li>
		</ul>
		</td>
		<td>
		<ul>
			<li>Delete gaps bug fixed.</li>
			<li>FileChooser sorts columns.</li>
			<li>Can remove groups one by one.</li>
			<li>Filechooser icons installed.</li>
			<li>Finder ignores return character when searching. Return key
			will initiate a search.<br>
			</li>
		</ul>
		</td>
	</tr>
	<tr>
		<td>
		<div align="center"><strong>2.0</strong><br>
		20/6/05</div>
		</td>
		<td>
		<ul>
			<li>New codebase</li>
		</ul>
		</td>
		<td>&nbsp;</td>
	</tr>
</table>
<p>&nbsp;</p>
</body>
</html>
