<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (Version 2.9)
 * Copyright (C) 2015 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<head>
<title>What's new ?</title>
</head>
<body>
  <p>
    <strong>What's new ?</strong>
  </p>
  <p>
    Jalview 2.9 has been in development since December 2014. In addition
    to a multitude of bug fixes and minor improvements (both small, and
    rather big!), it also brings major new capabilities for codon-level
    analysis of protein alignments and the retrieval and manipulation of
    structural data.</p><p>For the full list of changes, see the
    <a href="releases.html#Jalview.2.9">Jalview 2.9 Release Notes</a>.
  </p>
  <p>
    <strong>Highlights in Jalview 2.9</strong>
  <ul>
    <li><strong>Visualisation, editing and analysis of
        cDNA and Protein alignments</strong><br />A new <a
      href="features/splitView.html">Split View</a> window allows linked
      protein and nucleotide sequence alignments to be viewed, edited,
      and analysed as one. <br />cDNA alignments can also be
      reconstructed from protein alignments calculated by Jalview's web
      services, and update in response to edits in the amino acid view.<br />To
      start experimenting with cDNA/Protein analysis, jut drop a file
      containing cDNA sequences which code for proteins in an existing
      alignment, and Jalview will do the rest.</li>
    <li><strong>Enhanced Integration of UCSF Chimera</strong> <br>Jalview
      2.9 provides full support for the use of Chimera to view 3D
      structures linked to alignment views in the Jalview Desktop. We've
      also included support for saving Chimera sessions in Jalview
      project files.<br />Jalview and Chimera communicate using local
      web server connections, which may cause firewall alerts on some
      systems, but has the advantage of allowing bidirectional
      communication. Communication between Jalview and Chimera is now
      much more responsive, and selected regions in Chimera are now
      shown as highlighted regions in the Jalview desktop.</li>
    <li><strong>Interactive querying of the PDBe</strong><br />Jalview
      users can now <a href="features/pdbsequencefetcher.html">browse</a> and <a href="features/viewingpdbs.html">retrieve 3D structure</a> data from the PDB
      via the <a href="http://www.ebi.ac.uk/pdbe/api/doc/search.html">PDBe
        Search API</a> (<a href="http://dx.doi.org/10.1093%2Fnar%2Fgkt1180">Gutmanas
        et al 2014</a>). Developed in collaboration with the PDBe group at
      EMBL-EBI, the interface allows both structured and free-text
      queries to be performed, and allows automatic selection of the
      most relevant structures for an alignment acording to a variety of
      criteria.</li>
    <li><strong>Improved support for RNA visualisation</strong><br />Jalview
      2.9 integrates the latest version of the <a
      href="features/varna.html">VARNA RNA Viewer</a>, and VARNA views
      can also now be stored in Jalview projects. We've also dealt with
      a number of lingering bugs in the VARNA/Jalview interface,
      including the loss of pseudoknots when RNA secondary structure is
      shown VARNA.</li>
    <li><strong>Protein Secondary Structure predictions
        with JPred4</strong><br />Jalview includes a number of new features for
      working with secondary structure predictions from the JPred4
      server. These include new <a href="menus/popupMenu.html#hideinserts">popup menu actions</a> to automatically hide insertions and highlight
      mutations in an alignment with respect to a <a href="calculations/referenceseq.html">Reference
        Sequence</a>. Jalview 2.9's new <a href="io/export.html#htmlexport">scrollable
        SVG HTML export</a> was also developed specifically for the JPred4
      server.</li>
  </ul>

</body>
</html>
