<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (Version 2.7)
 * Copyright (C) 2011 J Procter, AM Waterhouse, G Barton, M Clamp, S Searle
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * 
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
-->
<head><title>Alignment Conservation Annotation</title></head>
<body><p><strong>Alignment Conservation Annotation</strong></p>
<p>This is an automatically calculated quantitative alignment
annotation which measures the number of conserved physico-chemical
properties conserved for each column of the alignment. Its calculation
is based on the one used in
  the AMAS method of multiple sequence alignment analysis :<br>
<ul>Livingstone
  C.D. and Barton G.J. (1993), Protein Sequence Alignments: A Strategy 
  for the Hierarchical Analysis of Residue Conservation.<em>CABIOS</em> Vol. <b>9</b>
  No. 6 (745-756)).
</ul>
</p>
<p>Conservation is measured as a numerical index reflecting the conservation of 
  <a href="../misc/aaproperties.html">physico-chemical 
  properties</a> in the alignment: Identities score highest, and the next most 
  conserved group contain substitutions to amino acids lying in the same physico-chemical 
  class.</p>

<p><em>Colouring an alignment by conservation</em><br>
Conservation scores can be used to colour an alignment.  This is
explained further in the help page for <a
href="../colourSchemes/conservation.html">conservation colouring</a>.
</p>
</body>
</html>
